Added video processing section (Unit 7 - Transformers based models)

mreraser opened the pull request:
Co-authored-by: seoulsky-field
<seoulsky.field02@gmail.com>

# What does this PR do?

Added `Transformers based models`  at video processing section. This document provides an overview of how Transformer models are applied in video processing, focusing on the Vision Transformer (ViT) and its video-specific variant, the Video Vision Transformer (ViViT), and TimeSFormer model. 

Thank you in advance for your review. 

Part of [Proposed Outline Revision for Unit 7. Video & Video Processing / dicussions #348](https://github.com/johko/computer-vision-course/discussions/348)

# Who can review?
@jungnerd @cjfghk5697 @1kmmk1 and anyone who wants to review!

# Who can review (Final)

jungnerd commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+
+## Introduction[[introduction]]
+
+In this chapter, we will cover how the Transformers model is utilized in video Processing. In particular, we will introduce the Vision Transformer, a successful application of the Transformers model in the field of Vision. We will then explain the additional considerations made for the Video Vision Transformer (ViViT) model used in video, as opposed to the Vision Transformer model used in images. ...
```
```suggestion
In this chapter, we will cover how the Transformers model is utilized in video processing. In particular, we will introduce the Vision Transformer, a successful application of the Transformers model in the field of vision. We will then explain the additional considerations made for the Video Vision Transformer (ViViT) model used in video, as opposed to the Vision Transformer model used in images. Finally, we will briefly discuss about the TimeSFormer model. 
```

jungnerd commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+
+- The model is pre-trained on large-scale datasets and fine-tuned for downstream vision tasks, achieving high performance.
+
+### Performance & Limitation[[performance-limitation]]
```
```suggestion
### Performance & Limitation[[performance-and-limitation]]
```

jungnerd reviewed (commented):
Fixed minor typos and suggested the name of the anchor. 
Other than that, everything looks good to me 👍🏻

jungnerd commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+</div>
+<small>Factorised self-attention (Model 3). Taken from the <a href = "https://arxiv.org/abs/2103.15691">original paper</a>.</small>
+
+In model 3, instead of computing multi-headed self-attention across all pairs of tokens, first only compute self-attention spatially (among all tokens extracted from the same temporal index). Then compute self-attention temporally(among all tokens extracted from the same spatial index). Because of the ambiguities no CLS(classification) token is used.
```
```suggestion
In model 3, instead of computing multi-headed self-attention across all pairs of tokens, first only compute self-attention spatially(among all tokens extracted from the same temporal index). Then compute self-attention temporally(among all tokens extracted from the same spatial index). Because of the ambiguities, no CLS(classification) token is used.
```

jungnerd commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+</div>
+<small>Factorised encoder (Model 2). Taken from the <a href = "https://arxiv.org/abs/2103.15691">original paper</a>.</small>
+
+First, only spatial interactions are contextualized through Spatial Transformer Encoder (=ViT). Then, each frame is encoded to a single embedding, fed into the Temporal Transformer Encoder(=general transformer).
```
```suggestion
First, only spatial interactions are contextualized through Spatial Transformer Encoder(=ViT). Then, each frame is encoded to a single embedding, fed into the Temporal Transformer Encoder(=general transformer).
```

jungnerd commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+
+After comparing Models 1, 2, 3, and 4, it is evident that Model 1 achieved the best performance but required the longest training time. In contrast, Model 2 demonstrated relatively high performance with shorter training times compared to Models 3 and 4, making it the most efficient model overall.
+
+ The ViViT model fundamentally faces the issue of dataset sparsity. Like the Vision Transformer (ViT), ViViT requires an extremely large dataset to achieve good performance. However, such a scale of dataset is often unavailable for videos. Given that the learning task is more complex, the approach is to first pre-train on a large image dataset using ViT to initialize the model. 
```
```suggestion
 The ViViT model fundamentally faces the issue of dataset sparsity. Like the Vision Transformer(ViT), ViViT requires an extremely large dataset to achieve good performance. However, such a scale of dataset is often unavailable for videos. Given that the learning task is more complex, the approach is to first pre-train on a large image dataset using ViT to initialize the model. 
```

johko commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+### Transformer models for video in ViViT[[transformer-models-for-video-in-vivit]]
+#### Model 1 : Spatio-temporal attention[[model-1-spatio-temporal-attention]]
+
+Simply forwards all spatio-temporal tokens extracted from the video, through the transformer encoder. Each frame is split n_w x n_h image patches, so total n_t x n_w x n_h patches contextualize from each other using transformer encoder. (n_h : # of rows, n_w : # of columns, n_t : # of frames)
```
I think this part needs to be a bit longer. 
- First of all I wonder what spatio-temporal tokens are and where they come from. 
- Then I think it would be nice to explain the "contextualize" part a bit more - what do you mean by that?
- And the part that is in brackets at the end actually really helps in understanding the things before, so maybe just use this and explain it before giving the "n_w x n_h image patches, so total n_t x n_w x n_h" part

johko commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+Simply forwards all spatio-temporal tokens extracted from the video, through the transformer encoder. Each frame is split n_w x n_h image patches, so total n_t x n_w x n_h patches contextualize from each other using transformer encoder. (n_h : # of rows, n_w : # of columns, n_t : # of frames)
+
+**complexity : O(n_h^2 x n_w^2 x n_t^2)**
+#### Model 2 : Factorised encoder[[model-2-factorised-encoder]]
```
```suggestion

#### Model 2 : Factorised Encoder[[model-2-factorised-encoder]]
```

johko commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+</div>
+<small>Factorised encoder (Model 2). Taken from the <a href = "https://arxiv.org/abs/2103.15691">original paper</a>.</small>
+
+First, only spatial interactions are contextualized through Spatial Transformer Encoder(=ViT). Then, each frame is encoded to a single embedding, fed into the Temporal Transformer Encoder(=general transformer).
```
```suggestion
First, only spatial interactions are contextualized through a Spatial Transformer Encoder(=ViT). Then, each frame is encoded to a single embedding and fed into the Temporal Transformer Encoder(=general transformer).
```

johko commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+
+**complexity : O(n_h^2 x n_w^2 + n_t^2)**
+
+#### Model 3 : Factorised self-attention[[model-3-factorised-self-attention]]
```
```suggestion
#### Model 3 : Factorised Self-Attention[[model-3-factorised-self-attention]]
```

johko commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_7_vivit_model3.JPG" alt="ViViT model 3"></img>
+</div>
+<small>Factorised self-attention (Model 3). Taken from the <a href = "https://arxiv.org/abs/2103.15691">original paper</a>.</small>
```
```suggestion
<small>Factorised Self-Attention (Model 3). Taken from the <a href = "https://arxiv.org/abs/2103.15691">original paper</a>.</small>
```

johko commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+</div>
+<small>Factorised self-attention (Model 3). Taken from the <a href = "https://arxiv.org/abs/2103.15691">original paper</a>.</small>
+
+In model 3, instead of computing multi-headed self-attention across all pairs of tokens, first only compute self-attention spatially(among all tokens extracted from the same temporal index). Then compute self-attention temporally(among all tokens extracted from the same spatial index). Because of the ambiguities, no CLS(classification) token is used.
```
```suggestion
In model 3, instead of computing multi-headed self-attention across all pairs of tokens, we first only compute self-attention spatially (among all tokens extracted from the same temporal index). Next, we compute self-attention temporally (among all tokens extracted from the same spatial index). Because of the ambiguities, no CLS (classification) token is used.
```

johko commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+In model 3, instead of computing multi-headed self-attention across all pairs of tokens, first only compute self-attention spatially(among all tokens extracted from the same temporal index). Then compute self-attention temporally(among all tokens extracted from the same spatial index). Because of the ambiguities, no CLS(classification) token is used.
+
+**complexity : same as model 2**
+#### Model 4 : Factorized dot-product attention[[model-4-factorized-dot-product-attention]]
```
```suggestion

#### Model 4 : Factorized Dot-Product Attention[[model-4-factorized-dot-product-attention]]
```

johko commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_8_vivit_model4.JPG" alt="ViViT model 4"></img>
+</div>
+<small>Factorised dot-product attention (Model 4). Taken from the <a href = "https://arxiv.org/abs/2103.15691">original paper</a>.</small>
```
```suggestion
<small>Factorised Dot-Product Attention (Model 4). Taken from the <a href = "https://arxiv.org/abs/2103.15691">original paper</a>.</small>
```

johko commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+In model 4, half of the attention heads are designed to operate with keys and values from spatial indices, the other half operate with keys and values from same temporal indices.
+
+**complexity : same as model 2, 3**
+### Experiments and Discussion[[experiments-and-discussion]]
```
```suggestion

### Experiments and Discussion[[experiments-and-discussion]]
```

johko replied:
Oh and thanks of course @mreraser , totally forgot to include you in the comment above ;)

johko reviewed (commented):
Thanks for the contribution @jungnerd !
I think the main base for most my comments is that we need to keep in mind that the course should also be good to read for beginners. I think sometimes you assume a bit too much prior knowledge, adding some more background info here and there would already be really great.

But apart from that it is a great piece of education, I already learned quite a few things just from reading through it once. Thank you so much for the effort :hugs:

mreraser commented:
@johko Thank you very much for your review! 😄 I will carefully read through the details you provided and make the necessary revisions accordingly.

mreraser commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+# Transformers in Video Processing (Part 1)[[transformers-in-video-processing-part-1]]
```
OK! I will remove all anchors!

mreraser commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,153 @@
+
+*Inspired by the Transformer scaling successes in NLP, we experiment with applying a standard Transformer directly to images, with the fewest possible modifications. To do so, we split an image into patches and provide the sequence of linear embeddings of these patches as an input to a Transformer. Image patches are treated the same way as tokens (words) in an NLP application. We train the model  ...
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_1_vit_architecture.png" alt="Vision transformer architecture"></img>
+</div>
+<small>ViT architecture. Taken from the <a href= "https://arxiv.org/abs/2010.11929"> original paper</a>.</small>
```
Yes I understand, and I’ll make adjustments if any issues arise in the future.

mreraser commented:
Hello @johko! 😃 

I have carefully reviewed your feedback and addressed the points you mentioned as follows:

1. Removed all anchor points
2. Fixed minor typos
3. Added an explanation of embeddings and their importance
4. Clarified the reason for discussing Uniform Frame Sampling and Tubelet Embeddings
5. Provided a definition of spatio-temporal tokens
6. Explained the term "contextualize"
7. Defined `n_w`, `n_h`, and `n_t` earlier in the text

Thank you for your guidance, and please let me know if there’s anything else you’d like me to improve!

johko reviewed (approved):
Thank you for the changes :slightly_smiling_face: 
LGTM :+1:

mreraser commented:
> Thank you for the changes 🙂 LGTM 👍

Thank you 👍 I also resolved some toctree conflits.

mreraser commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,178 @@
+*Inspired by the Transformer scaling successes in NLP, we experiment with applying a standard Transformer directly to images, with the fewest possible modifications. To do so, we split an image into patches and provide the sequence of linear embeddings of these patches as an input to a Transformer. Image patches are treated the same way as tokens (words) in an NLP application. We train the model  ...
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_1_vit_architecture.png" alt="Vision transformer architecture"></img>
```
```suggestion
<img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_1_vit_architecture.png" alt="Vision transformer architecture">
```

mreraser commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,178 @@
+### Performance & Limitation
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_2_vit_performance.JPG" alt="Vision transformer performance"></img>
```
```suggestion
<img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_2_vit_performance.JPG" alt="Vision transformer performance">
```

mreraser commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,178 @@
+*We present pure-transformer based models for video classification, drawing upon the recent success of such models in image classification. Our model extracts spatiotemporal tokens from the input video, which are then encoded by a series of transformer layers. In order to handle the long sequences of tokens encountered in video, we propose several, efficient variants of our model which factorise  ...
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_3_vivit_architecture.png" alt="ViViT architecture"></img>
```
```suggestion
<img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_3_vivit_architecture.png" alt="ViViT architecture">
```

mreraser commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,178 @@
+#### Tubelet Embedding
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_5_tubelet_embedding.JPG" alt="Tubelet embedding"></img>
```
```suggestion
<img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_5_tubelet_embedding.JPG" alt="Tubelet embedding">
```

mreraser commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,178 @@
+The approach in Model 1 was somewhat inefficient, as it contextualized all patches simultaneously. To improve upon this, Model 2 separates the spatial and temporal encoders sequentially.
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_6_vivit_model2.JPG" alt="ViViT model 2"></img>
```
```suggestion
<img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_6_vivit_model2.JPG" alt="ViViT model 2">
```

mreraser commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,178 @@
+#### Model 3 : Factorised Self-Attention
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_7_vivit_model3.JPG" alt="ViViT model 3"></img>
```
```suggestion
<img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_7_vivit_model3.JPG" alt="ViViT model 3">
```

mreraser commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,178 @@
+#### Model 4 : Factorized dot-product attention
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_8_vivit_model4.JPG" alt="ViViT model 4"></img>
```
```suggestion
<img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_8_vivit_model4.JPG" alt="ViViT model 4">
```

mreraser commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,178 @@
+### Experiments and Discussion
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_9_vivit_performance.JPG" alt="ViViT model performance"></img>
```
```suggestion
<img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_9_vivit_performance.JPG" alt="ViViT model performance">
```

mreraser commented on chapters/en/unit7/video-processing/transformers-based-models.mdx:
```diff
@@ -0,0 +1,178 @@
+TimeSFormer is a concurrent work with ViViT, applying Transformer on video classification. The following sections are explanations of each type of attention.
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_10_timesformer.JPG" alt="TimeSFormer model"></img>
```
```suggestion
<img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/transformer_based_video_model/unit7_10_timesformer.JPG" alt="TimeSFormer model">
```

ATaylorAerospace reviewed (approved):
Great additions..LGTM!
